keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,143 @@
+
+Currently, service portability as declared through Kubernetes with a common API is not possible. This feature set would provide a means to solve that.
+
+### Goals
+
+* Provide a common way to request common services (e.g., MySQL)
+* Provide a common means to obtain credentials to use the service
+* Provide a common method to detect which services are available
+* Provide a system that can be implemented for the major public clouds, on-premise clusters, and local cluster (e.g., Docker for Mac)
+
+### Non-Goals
+
+* Provide an out of the box solution for every bespoke service provided by everyone
+* Replace Kubernetes Service Catalog
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+#### Story 1
+
+As a user of Kubernetes, I can query the services I can declaratively reqest using a Kubernetes native API. For example, using the command `kubectl get crds` where I can see services alongside other resources that can be created.
+
+#### Story 2
+
+As a user of Kubernetes, I can declaratively request an instance of a service using a custom resource. When the service is provided the means to use that service (e.g., credentials in a secret) are provided in a common and consistent manner. The same resource and secret can be used in clusters running in different locations and the way the service is provided may be different.
+
+### Implementation Details/Notes/Constraints
+
+There are two important types of resources in this model. First, there are custom resource definitions (CRDs) for the services. Each service will have its own CRD. For example, if MySQL, PostgreSQL, and DNS are all available there will be three different CRDs that can be listed. To create an insteance of one a custom resource will be created following the API from the custom resource definition (CRD).
+
+The second type of resource is the optional secret to access the resource. If we continue the example, MySQL and PostgreSQL will have connection credentials that need to be used and injected into an appliction. In each case there will be a secret with a type specific to the service that follows a common schema. For services that do not require credential no secret is required.
+
+The CRDs and secret schemas will be published as an API that can be implemented by one or more controllers in the ecosystem.
+
+To operate on the CRDs and generate the appropriate instances and secrets this project will provide a controller that operates the services in Kubernetes clusters. This can be used Minikube, Docker for Mac Kubernetes, and other clusters. Controllers that leverage services native to cloud providers (e.g, AWS, Azure, and Google) can be implemented in other locations. This project will be happy to list 3rd party controllers but not host them or provide development or support of them.
```

Comment:
What this controller will do?

Reply:
As far as I understand this implies a "test/fake controller" that will only be suitable for testing on locally running Kubernetes (Minikube etc), not suitable for production.

Reply:
The controller built here would be scoped to that needed for testing and that needed to support Minikube (though I could see it being used in Docker for Mac). We will not support what's needed to handle all the varying cloud providers.

The goal is to foster interoperability and not to build a production system. Those are different goals.

This does open the ability for anyone to create controllers for cloud providers and not just those working for the cloud providers themselves. For example, ACME, Inc could create an AWS controller and share it for people to use. As could the company I work for. There is no requirement for just one implementation for a given cloud provider.